[RDY] UI grid protocol revision: line based updates and semantic highlights

bfredl opened the pull request:
Some changes I think it will make sense to group together as a new revision of the "grid protocol". This refers mostly to the `ui-grid` section of `ui.txt`, other parts of the ui protocol is not expected to change much.

This branch is currently a mess, but I prefer presenting messy code than vaporwave... It will be cleaned up and test added etc after some discussion this is the direction we want to move in.
 
For backwards compatibility `api/ui.c` should translate back to the present protocol, unless clients opt-in. All internal structures such as bridge, TUI, the "compositor" in #6619 should use the new format. 

- [x] For clients that opt-in to this revision, we can remove deprecated events, such as `update_fg`, `eol_clear` etc.
- [x] Nominally support multiple grids. Enabling this revision shouldn't _per-se_ cause external clients to see any more grids. But grid events should be updated as needed for multi-grid support. Multi-grid is an abstraction which can be used by several concrete features (where floating windows which will have its own `ext_float` flag, will be the first)
- [x] support sending an entire updated screen line at a time. Or rather, the largest damaged region of each line. Also supports clearing to an arbitrary column with non-normal background (and thus replaces severely limited `eol_clear`).
- [x] define highlight groups in advance, and later just send numeric ids. Always sends both rgb and cterm colors, `rgb` flag thus has no effect.
- [x] (Separate) opt in: allow UIs to inspect the "semantic" highlight state. UI highlights like `Number` will be identified as such, and UIs need not guess it from color values or positions. (current representation is very detailed, we might not want this level of detail, but I wanted to see how far it could go...)
- [x] But if the UI doesn't opt in, it should only see highlight ids with unique colors (should also apply to TUI, to reduce spurious updates)
- [x] Changing the `Normal` color shouldn't cause clearing the screen. The UI already has all necessary information to repaint the screen. Not a very impactful optimization as such, but this convention will be used later to support efficient highlighting overlays.

It probably makes sense to mark this revision as experimental/unstable in the first PR that introduces it and use some unwieldy name as `ext_multigrid_experimental`, and recommend most external UIs to stick with the present format for now. A good reference point for stabilizing it is the same time as merging the floats PR #6619, as then there will be a strong reason for clients to actually use it. We should see this an opportunity to fix the issues/limitations of the UI protocol that has accumulated over time. Moving forward we only need to support the current format, and this new revision when we are happy with it. 

Depends on: #7992

bfredl commented:
Ref #7723 (pre-registered colors)
#6459 (more efficient line representation)
#8006 (forward highlight groups)

bfredl committed:
highlight: extract low-level highlight logic from syntax, ui

bfredl committed:
highlight: refactor to use stateful representation

This allows us to keep track of the source higlight groups,
and not only the final combined highlights.

bfredl commented:
Cleaned up the branch a little.

Concretely, here are a sketch new and changed events in the revised grid protocol. Names and such are meant to be bikeshedded (and not 100% consistent with the branches either).

`hl_attr_define(integer id, Dictionary attr, Array info)`
Tell UI that than a highlight with `id` was added to the highlight table. The `attr` dict has similar format to the old `highlight_set`, except that both rgb and cterm colors are always available. `info` is empty by default.

`hl_attr_set(integer id)`
Works like `highlight_set`, except clients should look up the `id` in an earlier `hl_attr_define` event.

`grid_line(Integer grid_id, Integer row, Integer col_start, Integer col_end, Array cells)`
Tell the UI that the chunk `[col_start, col_end[` of `row` on grid `grid_id` was changed. cells in an array of `[text, hl_id]` each describing the UTF-8 text and highlight id (as above) of this cells. Right cells of doublewidth chars will be the emtpy string (unlike "empty" cells, which contains space char). If the cells doesn't reach all the way to `col_end`, the rest of the chunk should be cleared with the current highlight.

As an alternative, we could allow cell tuples to optionally be `[text, hl_id, repeat]`. Then they could do efficient filling even of non-blanks, and `col_end` can be dropped.

`grid_resize(grid_id, rows, cols)` (replacing resize)
Resize grid with `grid_id`. If `grid_id` wasn't seen by the client before, a new grid is being created with this size. Grid 1 is the "global" grid used by default. The purpose of grids beyond that will be defined by further extensions/events. Thus, just activating `ext_multigrid` will not cause the client to "see" any grids it cannot handle.

`grid_destroy(grid_id)`
`grid_id` will not be used anymore and the client can free all data associated with it.

`grid_cursor_goto(grid_id, row, column)` (replacing `cursor_goto`)
Makes grid `grid_id` current and `row,column` the cursor position on that event. scrolling will be done on the current grid. The grid that is active at the end of a `redraw` batch is considered the focused grid (this is consistent with how the cursor is managed already)

The following events are redundant with these additions and will not be sent in the new mode:
`eol_clear` (`line` can do this more generally)
`update_fg` (use `default_colors_set`)
`update_bg`
`update_sp`
`highlight_set`

justinmk commented:
Would this also be the right time to separate the "pen" from the "edit cursor", i.e. disambiguate `cursor_goto`.

bfredl commented:
@justinmk That would be done by never using `ui_put` and always use `ui_line`. Currently, use of `ui_put` means code that haven't been refactored yet.

The lazyredraw-y approach would be to keep track of largest changed region of each line, and transmit that only at `ui_flush`. That could be messy with scrolling though. I will first try to convert most stuff to `ui_line` in a "semantic" way. (which also will need keep track of dirty regions, but only for a single line in a well defined context, and not globally)

bfredl commented:
Though then `scroll` will still depend on the cursor grid, but we could change it to `grid_scroll(grid_id, row, col, width, height, scroll_rows, scroll_cols)` as scrolling region is always defined prior to scroll anyway.

bfredl committed:
ui: add TODO for non-working terminal linewrap

Note: this has not been working since the TUI reimplementation

bfredl commented:
@justinmk This is now 85% done. `cursor_goto` is _only_ sent once at the end of a batch, and indicates the final position. The remaining 15% is to make sure the cursor position is actually correct, especially with messages as cursor position is almost never set explicitly but rather implicitly by the last put char. Thankfully we happen to have a large body of screen tests checking the indented position...

In addition, `ui_put` and `ui_set_highlight` are  _removed_ from the internal protocol as they are now redundant. `api/ui.c` is still emulating these events for existing remote UIs, which includes the present `screen.lua` implementation (which thus serves as test for the translation).

bfredl commented:
Now stateful cursor should work. Though one place it works slightly different is with ext_cmdline, it now didn't move the cursor to the cmdline row in some cases, but I think this is an actual improvement. I'd try make sure this new behavior is consistent, even with redrawing screen in cmdline mod.

At least `functional/ui` tests now pass (with the change of ext_cmdline behavior)

bfredl commented:
Also the behavior of `writedelay` is changed, the delay is now per line update, which at least makes sense while developing this branch. The same UI component should ideally only use one line update per line. At least window line, statusline-ish things and pum now follow this.

Messages are more tricky, as messages do treat the screen as a "dumb" terminal semantically. But I think we can live with that for now. (it is still not one line update per char, more like one per logical item)

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -1562,3 +1563,15 @@ Object nvim_get_proc(Integer pid, Error *err)
   return rvobj;
 }
+
+Array nvim__inspect(Integer row, Integer col)
```
though this is "private", it could still benefit from a more intuitive name, maybe `nvim__screen_inspect` or `nvim__ui_inspect`

bfredl replied:
Currently it is mostly for my own debugging, it will be made redundant or replaced with something better.

bfredl commented:
Scrolling is now stateless. Stateful scrolling regions was never actually used for any good, and more an arbitrary artefact of how the code in `screen.c` was organized.

bfredl commented:
The set of events of the new grid protocol are now consistently named and disjoint from the old event set (if the event is shared, it is not considered a grid event).

Also implemented the new protocol in the test suite, while of course still supporting the old. (modulo one remaining failing test in `ui/`, haven't checked the rest). In principle we could run some part of the test suite (almost all of `test/functional/ui` ?) twice, once for each revision, but not sure how to best do that.

oni-link commented on src/nvim/api/ui.c:
```diff
@@ -241,27 +242,282 @@ static void push_call(UI *ui, char *name, Array args)
+    if (endcol < clearcol) {
+      remote_ui_highlight_set(ui, (int)clearattr);
+      if (clearattr == 0 &&  clearcol == Columns) {
+        remote_ui_cursor_goto(ui, row, endcol);
```
This line could be move in front of the if-block (see also other branch).

bfredl committed:
ui: docs for ext_newgrid and ext_hlstate

bfredl commented:
I began sketching on actual documentation (see last commit). Not really happy with `grid_line` (the core event of the new protocol) doc yet, suggestions welcome.

UtkarshMe commented:
Hi, I was reading through the pull request and found some typos. It's not much helpful but here's a patch for it.
```patch
From 2a7ab76d30c7905ed79e67df249c757615a064db Mon Sep 17 00:00:00 2001
From: UTkarsh Maheshwari <utkarshme96@gmail.com>
Date: Wed, 2 May 2018 23:16:50 +0530
Subject: [PATCH] Fix typos

---
 runtime/doc/ui.txt        | 2 +-
 src/nvim/api/ui.c         | 2 +-
 src/nvim/highlight.c      | 8 ++++----
 src/nvim/highlight_defs.h | 2 +-
 4 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/runtime/doc/ui.txt b/runtime/doc/ui.txt
index 225970acb..094eddc4b 100644
--- a/runtime/doc/ui.txt
+++ b/runtime/doc/ui.txt
@@ -210,7 +210,7 @@ option is not set).
 ["set_scroll_region", top, bot, left, right]
        Define the scroll region used by `scroll` below.
 
-       Note: ranges are end-incluvise, which is inconsistent with API
+       Note: ranges are end-inclusive, which is inconsistent with API
        conventions.
 
 ["scroll", count]
diff --git a/src/nvim/api/ui.c b/src/nvim/api/ui.c
index dbe337f07..62fe91e7e 100644
--- a/src/nvim/api/ui.c
+++ b/src/nvim/api/ui.c
@@ -207,7 +207,7 @@ static void ui_set_option(UI *ui, bool init, String name, Object value,
       bool boolval = value.data.boolean;
       if (!init && i == kUIMultigrid && boolval != ui->ui_ext[i]) {
         // There shouldn't be a reason for an UI to do this ever
-        // so expclicitly not support it.
+        // so explicitly not support it.
         api_set_error(error, kErrorTypeValidation,
                       "ext_multigrid option cannot be changed");
       }
diff --git a/src/nvim/highlight.c b/src/nvim/highlight.c
index b9c29973b..bbdcc89d5 100644
--- a/src/nvim/highlight.c
+++ b/src/nvim/highlight.c
@@ -49,7 +49,7 @@ void highlight_use_hlstate(void)
 static int get_attr_entry(HlEntry entry)
 {
   if (!hlstate_active) {
-    // This infromation will not be used, erase it and reduce the table size.
+    // This information will not be used, erase it and reduce the table size.
     entry.kind = kHlUnknown;
     entry.id1 = 0;
     entry.id2 = 0;
@@ -76,7 +76,7 @@ static int get_attr_entry(HlEntry entry)
     highlight_attr_set_all();
 
     recursive = false;
-    if (entry.kind == kHlComibne) {
+    if (entry.kind == kHlCombine) {
       // This entry is now invalid, don't put it
       return 0;
     }
@@ -260,7 +260,7 @@ int hl_combine_attr(int char_attr, int prim_attr)
     }
   }
 
-  id = get_attr_entry((HlEntry){ .attr = new_en, .kind = kHlComibne,
+  id = get_attr_entry((HlEntry){ .attr = new_en, .kind = kHlCombine,
                                  .id1 = char_attr, .id2 = prim_attr });
   if (id > 0) {
     map_put(int, int)(combine_attr_entries, combine_tag, id);
@@ -406,7 +406,7 @@ void hl_inspect_impl(Array *arr, int attr)
       PUT(item, "kind", STRING_OBJ(cstr_to_string("term")));
       break;
 
-    case kHlComibne:
+    case kHlCombine:
       hl_inspect_impl(arr, e.id1);
       hl_inspect_impl(arr, e.id2);
       return;
diff --git a/src/nvim/highlight_defs.h b/src/nvim/highlight_defs.h
index 1499800cd..5f637228b 100644
--- a/src/nvim/highlight_defs.h
+++ b/src/nvim/highlight_defs.h
@@ -157,7 +157,7 @@ typedef enum {
   kHlUI,
   kHlSyntax,
   kHlTerminal,
-  kHlComibne,
+  kHlCombine,
 } HlKind;
 
 typedef struct {
-- 
2.17.0

```

bfredl commented:
Update: highlight state now actually should be correct (before stupid off-by-one error in syntax.c)

It is possible to inspect the additional info that an UI can get from a screen cell by:
```
call nvim__inspect_cell(0,0)
function! Mouse()
    while v:true
        let c = getchar()
        if c == "\<LeftMouse>"
            echo nvim__inspect_cell(v:mouse_y, v:mouse_x)[2]
            break
        end
    endwhile
endfunction
map ,x <cmd>call Mouse()<cr>
```

(yes I had to invent `mouse_x` and `mouse_y`, sorry about that)

bfredl committed:
ui: add tests for hlstate extension

bfredl commented:
So, I think this is basically "done", except for a few rough edges and maybe a few more tests should be added, and docs can always be improved. Marking RFC.

justinmk commented on src/nvim/syntax.c:
```diff
@@ -6909,8 +6907,16 @@ void do_highlight(const char *line, const bool forceit, const bool init)
+        // changed
+        ui_refresh();
+      } else {
+        // TUI and never UIs will repaint the screen themselves. NOT_VALID
```
never=newer

breja commented on src/nvim/api/ui_events.in.h:
```diff
@@ -64,6 +39,49 @@ void set_icon(String icon)
+  FUNC_API_SINCE(4) FUNC_API_REMOTE_IMPL;
+void grid_cursor_goto(Integer grid, Integer row, Integer col)
+  FUNC_API_SINCE(4) FUNC_API_REMOTE_IMPL;
+void grid_line(Integer grid, Integer row, Integer startcol, Integer clearcol,
```
is this function signature outdated? i don't see `clearcol` in the `grid_line` event

bfredl replied:
Thanks, I will change it.

bfredl commented:
There is a slight issue with the new highlight representation (for external UI, TUI is fine): it assues only colors are different, but the bitfield attributes are also different between `cterm` and `rgb`. So we need to either

- change the `hl_attr_define` event to transmit two dictionaries. Simplest to adapt existing code. 2x bandwidth cost, but this is part of 100x bandwidth reduction anyway...
- change the attributes to be tuples of booleans.
- change to use bitfield representation. Simple and compact, but requires numerical constants as part of the protocol description.

I'm leaning towards the first one, as it is simplest to implement and to use by existing client code.

bfredl committed:
ui: use line-based rather than char-based updates in screen.c

Add ext_newgrid and ext_hlstate extensions. These use predefined
highlights and line-segment based updates, for efficiency and
simplicity.. The ext_hlstate extension in addition allows semantic
identification of builtin and syntax highlights.

Reimplement the old char-based updates in the remote UI layer, for
compatibility. For the moment, this is still the default. The bulitin
TUI uses the new line-based protocol.

cmdline uses curwin cursor position when ext_cmdline is active.

bfredl commented:
`hl_attr_define` now has signature `[id, rgb_attr, cterm_attr, info]` to fully support `cterm` attrs. Most GUIs can just store the `rgb_attr` dict and throw away `cterm_attr`.

bfredl commented:
All tests now pass in travis at least, both using the "newgrid" representation and the old representation (implemented by translation in `api/ui.c` ). An idea could be to use different representations in different test folds, maybe using the old one in ASAN and coverage builds as it exposes slightly more C code. (as hlstate tests will use newgrid regardless, both branches will be exposed for fundamental operations).

bfredl commented:
I will merge this quite soon if there is no protests. Note some optimizations are missing, such that  messages will still use multiple updates per line, and the docs can always be improved, but these are improvements that can be done in a later PRs, in parallel with floating windows and wincmd exernalisation.

oni-link commented:
If I switch between light and dark background (`set background={light,dark}`), only the background with text is changed. Scrolling in the window fixes it.
Similar effect when starting nvim. The text in the center of the buffer has different background color as the rest of the buffer.

bfredl commented:
@oni-link I cannot reproduce with `-u NONE`. What color scheme do you use?

oni-link commented:
init.vim:
```
call plug#begin('~/.vim/bundle')

" Colorschemes
Plug 'morhetz/gruvbox'
call plug#end()

colorscheme gruvbox
```
`nvim -u init.vim README.md` shows different background color for text and empty space. Also when switching background between `dark` and `light`.

oni-link commented:
@bfredl , yes the last changes fixed it for me too.

blueyed commented:
I am seeing a segfault together with Neomake due to this: https://github.com/neovim/neovim/commit/989b585e105b4f343f639abf5daf820de19d4a26#commitcomment-29838744

bfredl commented:
Yeah a null check is needed, tough one was needed even before this PR (for anyone clearing the cursorline syntax group). Or maybe it should just be refactored to per value already, there is nothing wrong with returning a struct.

aktau commented:
I think I'm seeing what @blueyed is seeing too. Here's my backtrace: https://gist.githubusercontent.com/aktau/8b97030d5385087ea5ee249d9082a2b5/raw/3d8595dbd8be585ef78019366880731419b085ab/backtrace.log. Which looks similar to @blueyed's backtrace, although mine is much (much) larger (because of what some script doest). I'm not sure which script is causing it. It's got something to do with split screens and signs/columns I think.

bfredl commented:
#8789 might fix it. I will try to reproduce tomorrow, at least something similar (`:sign` right after `hi clear Statusline` might be enough, but it might be a bit contrived)
